libraries/AC_AutoTune/AC_AutoTune_FreqResp.cpp

```diff
@@ -0,0 +1,448 @@
+*/
+
+#include <AP_HAL/AP_HAL.h>
+#include "AC_AutoTune_FreqResp.h"
+
+// Initialize the Frequency Response Object. Must be called before running dwell or frequency sweep tests
+void AC_AutoTune_FreqResp::init(InputType input_type)
+{
+    excitation = input_type;
+    max_target_cnt = 0;
+    min_target_cnt = 0;
+    max_meas_cnt = 0;
+    min_meas_cnt = 0;
+    input_start_time_ms = 0;
+    new_tgt_time_ms = 0;
+    new_meas_time_ms = 0;
+    new_target = false;
+    new_meas = false;
+    curr_test_freq = 0.0f;
+    curr_test_gain = 0.0f;
+    curr_test_phase = 0.0f;
+    max_accel = 0.0f;
+    max_meas_rate = 0.0f;
+    max_command = 0.0f;
+    meas_peak_info_buffer->clear();
+    tgt_peak_info_buffer->clear();
+    cycle_complete = false;
+}
+
+// update_rate - this function receives time history data during a dwell and frequency sweep tests for max gain,
+// rate P and rate D tuning and determines the gain and phase of the response to the input. For dwell tests once
+// the designated number of cycles are complete, the average of the gain and phase are determined over the last 5 cycles
+// and the cycle_complete flag is set. For frequency sweep tests, phase and gain are determined for every cycle and
+// cycle_complete flag is set to indicate when to pull the phase and gain data.  The flag is reset to enable the next
+// cycle to be analyzed.
+void AC_AutoTune_FreqResp::update_rate(float tgt_rate, float meas_rate, float tgt_freq)
+{
+    uint32_t now = AP_HAL::millis();
+    uint32_t half_cycle_time_ms = 0;
+    uint32_t cycle_time_ms = 0;
+
+    if (cycle_complete) {
+        return;
+    }
+
+    if (!is_zero(tgt_freq)) {
+        half_cycle_time_ms = (uint32_t)(300 * M_2PI / tgt_freq);
+        cycle_time_ms = (uint32_t)(1000 * M_2PI / tgt_freq);
+    }
+
+    if (input_start_time_ms == 0) {
+        input_start_time_ms = now;
+    }
+
+    // cycles are complete! determine gain and phase and exit
+    if (max_meas_cnt > AUTOTUNE_DWELL_CYCLES + 1 && max_target_cnt > AUTOTUNE_DWELL_CYCLES + 1 && excitation == DWELL) {
+        float delta_time = 0.0f;
+        float sum_gain = 0.0f;
+        uint8_t cnt = 0;
+        uint8_t gcnt = 0;
+        uint16_t meas_cnt, tgt_cnt;
+        float meas_ampl = 0.0f;
+        float tgt_ampl = 0.0f;
+        uint32_t meas_time = 0;
+        uint32_t tgt_time = 0;
+        for (int i = 0;  i < AUTOTUNE_DWELL_CYCLES; i++) {
```

Comment:
this "int" should probably be "uint8_t".

Reply:
done